display_list/display_list_builder.cc

```diff
@@ -753,33 +749,12 @@ void DisplayListBuilder::clipPath(const SkPath& path,
   switch (clip_op) {
     case SkClipOp::kIntersect:
       Push<ClipIntersectPathOp>(0, 1, path, is_aa);
-      if (!path.isInverseFillType()) {
-        intersect(path.getBounds());
-      }
       break;
     case SkClipOp::kDifference:
       Push<ClipDifferencePathOp>(0, 1, path, is_aa);
-      // Map "kDifference of inverse path" to "kIntersect of the original path".
-      if (path.isInverseFillType()) {
-        intersect(path.getBounds());
-      }
       break;
   }
-}
-void DisplayListBuilder::intersect(const SkRect& rect) {
-  SkRect dev_clip_bounds = getTransform().mapRect(rect);
-  if (!current_layer_->clip_bounds().intersect(dev_clip_bounds)) {
-    current_layer_->clip_bounds().setEmpty();
-  }
-}
-SkRect DisplayListBuilder::getLocalClipBounds() {
-  SkM44 inverse;
-  if (current_layer_->matrix().invert(&inverse)) {
```

Comment:
I noticed that Data4x4 inverts the 3x3 part rather than the entire 4x4 matrix as this code does. I wonder if that matters and if we should file a bug against Data4x4?

Reply:
Also, even if the math works out the same for the coordinates, would doing the invert on the full 4x4 fail in some circumstances that an invert of the 3x3 succeeds? Perhaps if the unused entries caused a zero determinant? Or if they were non-finite?

Reply:
I found that SkCanvas also converts M44 to M33 and then inverse, so maybe the code in tracker should be adopted here.
c.f.
https://github.com/google/skia/blob/3724b5964eed55f5297fd5101a077f4180bfdcf5/src/core/SkCanvas.cpp#L1730